### PR TITLE
fix: filtering client suppliers MUST check the client region after creation as well as before

### PR DESCRIPTION
**Pull request by mattsb42-aws**
*Issue #, if available:*
resolves: #258 

*Description of changes:*

Per #258, this introduces two changes to the filtering client suppliers:
1. They now accept `None` to let users filter out requests when the region is unknown
1. In addition to checking the requested region before calling down to the inner client supplier, they now also check the client returned by the client supplier before returning it. This is to accommodate the scenario where the desired region is unknown and is set in the client by the inner client supplier.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: test/functional/keyrings/aws_kms/test_client_suppliers.py [hunk 1]
```diff
@@ -114,3 +134,24 @@ def test_allow_deny_nested_supplier():
...
+
+
+def test_allow_region_block_on_response():
+    test = AllowRegionsClientSupplier(
+        allowed_regions=[None, "us-west-2"], client_supplier=AlwaysOneRegionClientSupplier(region_name="us-east-2")
+    )
+
+    # Catch the blocked region on response.
+    with pytest.raises(UnknownRegionError):
+        test(None)
```
Comment (acioc) [inline test/functional/keyrings/aws_kms/test_client_suppliers.py:157]: Should we also test the positive cases? Passing None but having it be a valid region once it's checked?

Comment (mattsb42-aws) [reply test/functional/keyrings/aws_kms/test_client_suppliers.py:181]: Good call. Additional tests added.

**Comment by mattsb42-aws** [comment]
Discarding this change due to upcoming major rework of AWS KMS keyrings.

https://github.com/awslabs/aws-encryption-sdk-specification/blob/master/proposals/2020-07-01_aws-kms-keyring-redesign/change.md
